test/Spec/BotPlutusInterface/Collateral.hs

```diff
@@ -0,0 +1,107 @@
+module Spec.BotPlutusInterface.Collateral where
+
+import BotPlutusInterface.Types (
+  ContractEnvironment (cePABConfig),
+  PABConfig (pcOwnPubKeyHash), CollateralUtxo (CollateralUtxo)
+ )
+import Cardano.Api (TxBodyContent (txIns))
+import Cardano.Api qualified as C
+import Control.Lens ((&), (.~))
+import Control.Monad (when)
+import Data.Default (def)
+import Data.Set qualified as Set
+import Data.Text (Text, pack)
+import Data.Text qualified as Text
+import Ledger (PaymentPubKeyHash (unPaymentPubKeyHash), Tx (txCollateral, txInputs), txInRef)
+import Ledger.Ada qualified as Ada
+import Ledger.Constraints qualified as Constraints
+import Ledger.Tx (CardanoTx, TxOut (TxOut), TxOutRef (TxOutRef))
+import Ledger.Tx.CardanoAPI qualified as C
+import Plutus.Contract (
+  Contract (..),
+  Endpoint,
+  submitTx,
+  throwError,
+  ownPaymentPubKeyHash,
+ )
+import Spec.MockContract (
+  contractEnv,
+  paymentPkh1,
+  paymentPkh2,
+  pkhAddr1,
+  runContractPure,
+  theCollateralUtxo,
+  utxos,
+ )
+import Plutus.Contract.Logging (logInfo)
+import Test.Tasty (TestTree, testGroup)
+import Test.Tasty.HUnit (Assertion, assertFailure, testCase)
+import Prelude
+
+tests :: TestTree
+tests =
+  testGroup
+    "Doesn't spend collateral."
+    [ testCase "Use collateral utxo as collateral and not as input" testTxUsesCollateralCorrectly
+    ]
+
+-- | check that tx doesn't spend collateral but uses it as collateral
+testTxUsesCollateralCorrectly :: Assertion
```

Comment:
The current test for collateral is not right, because it only sends ada from user1 to user2 for that collateral is not required.

 But the "strange" thing is `submitTx` always produces, the following Tx:
 ```Left (SomeTx (ShelleyTx ShelleyBasedEraAlonzo (ValidatedTx {body = TxBodyConstr TxBodyRaw {_inputs = fromList [TxInCompact (TxId {_unTxId = SafeHash "5d677265fa5bb21ce6d8c7502aca70b9316d10e958611f3c6b758f65ad959996"}) 1,TxInCompact (TxId {_unTxId = SafeHash "76ed2fcda860de2cbacd0f3a169058fa91eff47bc1e1e5b6d84497159fbc9300"}) 0,TxInCompact (TxId {_unTxId = SafeHash "9405c89393ba84b14bf8d3e7ed4788cc6e2257831943b58338bee8d37a3668fc"}) 0,TxInCompact (TxId {_unTxId = SafeHash "a1be9565ccac4a04d2b5bf0d0167196ae467da0d88161c9c827fbe76452b24ef"}) 0], _collateral = fromList [TxInCompact (TxId {_unTxId = SafeHash "76ed2fcda860de2cbacd0f3a169058fa91eff47bc1e1e5b6d84497159fbc9300"}) 0], _outputs = StrictSeq {fromStrict = fromList [(Addr Testnet (KeyHashObj (KeyHash "0f45aaf1b2959db6e5ff94dbb1f823bf257680c3c723ac2d49f97546")) StakeRefNull,Value 999081125 (fromList []),SNothing),(Addr Testnet (KeyHashObj (KeyHash "0f45aaf1b2959db6e5ff94dbb1f823bf257680c3c723ac2d49f97546")) StakeRefNull,Value 1379280 (fromList [(PolicyID {policyID = ScriptHash "1d6445ddeda578117f393848e685128f1e78ad0c4e48129c5964dc2e"},fromList [("testToken",853543)])]),SNothing),(Addr Testnet (KeyHashObj (KeyHash "6696936bb8ae24859d0c2e4d05584106601f58a5e9466282c8561b88")) StakeRefNull,Value 1379280 (fromList [(PolicyID {policyID = ScriptHash "1d6445ddeda578117f393848e685128f1e78ad0c4e48129c5964dc2e"},fromList [("testToken",18)])]),SNothing),(Addr Testnet (KeyHashObj (KeyHash "981fc565bcf0c95c0cfa6ee6693875b60d529d87ed7082e9bf03c6a4")) StakeRefNull,Value 1379280 (fromList [(PolicyID {policyID = ScriptHash "1d6445ddeda578117f393848e685128f1e78ad0c4e48129c5964dc2e"},fromList [("testToken",15)])]),SNothing)]}, _certs = StrictSeq {fromStrict = fromList []}, _wdrls = Wdrl {unWdrl = fromList []}, _txfee = Coin 204837, _vldt = ValidityInterval {invalidBefore = SNothing, invalidHereafter = SNothing}, _update = SNothing, _reqSignerHashes = fromList [KeyHash "0f45aaf1b2959db6e5ff94dbb1f823bf257680c3c723ac2d49f97546"], _mint = Value 0 (fromList []), _scriptIntegrityHash = SNothing, _adHash = SNothing, _txnetworkid = SNothing}, wits = TxWitnessRaw {_txwitsVKey = fromList [], _txwitsBoot = fromList [], _txscripts = fromList [], _txdats = TxDatsRaw (fromList []), _txrdmrs = RedeemersRaw (fromList [])}, isValid = IsValid True, auxiliaryData = SNothing})) AlonzoEraInCardanoMode)```

this `Tx` is completely different from the required `Tx`,  I currently don't know why this happens, but the correct way to validate the correctness of the contract seems to be using `assertContract`